networkx/generators/random_graphs.py

```diff
@@ -1293,3 +1294,47 @@ def my_function(b):
             j = int(math.ceil(n * kernel_root(i / n, j / n, r)))
             graph.add_edge(i - 1, j - 1)
     return graph
+
+
+@py_random_state(2)
+def random_ordered_tree(n, seed=None, directed=False):
+    """
+    Creates a random ordered tree
+
+    Parameters
+    ----------
+    n : int
+        A positive integer representing the number of nodes in the tree.
+
+    seed : integer, random_state, or None (default)
+        Indicator of random number generation state.
+        See :ref:`Randomness<randomness>`.
+
+    directed : bool
+        if the edges are one-way
+
+    Returns
+    -------
+    networkx.OrderedDiGraph | networkx.OrderedGraph
+
+    Example
+    -------
+    >>> import networkx as nx
+    >>> assert len(nx.random_ordered_tree(n=1, seed=0).nodes) == 1
+    >>> assert len(nx.random_ordered_tree(n=2, seed=0).nodes) == 2
+    >>> assert len(nx.random_ordered_tree(n=3, seed=0).nodes) == 3
+    >>> otree = nx.random_ordered_tree(n=5, seed=3, directed=True)
+    >>> print(nx.forest_str(otree))
+    ╙── 0
+        └─╼ 1
+            └─╼ 4
+                ├─╼ 2
+                └─╼ 3
+    """
+    from networkx.utils import create_py_random_state
+
+    rng = create_py_random_state(seed)
+    # Create a random undirected tree
+    create_using = nx.OrderedDiGraph if directed else nx.OrderedGraph
```

Comment:
You almost certainly don't need an `OrderedGraph` class here. That would be needed for Python version pre-3.6. But with Python3.6+ the regular classes are ordered because the underlying dict is ordered.

Reply:
This function is likely being removed as it is redundant with the new `create_using` arg in `random_tree`.

While dictionaries did becomes ordered by default in 3.6, that guarantee is still not part of the Python language spec AFAIK. I would much prefer to be explicit about the orderedness of graphs. Furthermore, there are mathematical implications of having an ordered graph versus one that has an arbitrary order. The algorithms in the PRs that build on this actually depend on the graph having a defined order for correctness and runtime analysis. I know there are randomized algorithms that depend on random orderings; I'm not sure if any rely on arbitrary orderings.

I'm stressing this to make sure the maintainers know there is value in explicitly using and keeping `Ordered` `Graph/DiGraph/etc...` variants in the API.

Reply:
Thanks for that comment.  
Also, CPython 3.7+ guarantees the order of dicts to be the order keys are added.  I guess if you try to use Pypy that is not guaranteed, but they are still on version2.7 and our current code probably is starting to break on that version.

Reply:
FWIW pypy has a 2.7 and a 3.6 version. I imagine order will be added to the Python dict spec in the future. My main concern is that networkx maintains the ability to distinguish between a graph that is ordered and a graph that has arbitrary order (even if that's not the implementation under the hood).

Reply:
We are only testing on PyPy 3.6 now.  I am waiting on this PR https://github.com/actions/setup-python/pull/168 to be merged and I will switch us to PyPy 3.7.  They have been working on getting the PyPy 3.7 support regularly for the last few weeks.

Reply:
We have only used ``OrderedGraph`` for testing in the past.  I intended to propose deprecating it once we drop Python 3.6, which will happen very soon.

Reply:
Be careful about claiming alignment between OrderedGraph and a (mathematical) ordered graph. I don't think it is accurate to use these classes to distinguish between "a graph that is ordered and a graph that has arbitrary order for algorithmic purposes". Certainly it is important to distinguish them mathematically. But both OrderedGraph and Graph have the same deterministic node and edge reporting order based on the order that nodes and edges were added to the object. 

We should deprecate OrderedGraph, though it might be useful to show an example of subclassing.

Reply:
My main concern is that the algorithm I implemented in #4327 explicitly requires an OrderedDiGraph, and I do [check for that](https://github.com/networkx/networkx/blob/9c4f3a59d5e273142d69b6f23b33b687384960b2/networkx/algorithms/minors/tree_embedding.py#L145). But this is just an input type check, and I suppose given that a DiGraph is ordered under the hood, the algorithm will still work. I'm worried that without this check a user might think that they are getting the maximum embedding of unordered graphs (which is an APX-hard problem, and the algo I wrote does not find that). I recognize that functionally they are equivalent. I just believe that the semantics are important for implementation elegance.

That's my main argument for keeping ordered graph variants, and if others don't think its strong enough to justify its existence, then I'll accept that.

Regardless, this function and all references to the OrderedDiGraph and OrderedGraph class no longer exist in this PR. So perhaps we can move forward with merging this and save the deprecation discussion for another issue.